protocol/builder-separation.md

```diff
@@ -0,0 +1,115 @@
+
+PBS is *only* the abstraction of block-building as an activity, which *allows for* blocks to be used in the system when they originate from somewhere other than the Execution Engine. The "somewhere else" is considered opaque for the purposes of this document, except when the its properties inform the design of the interface.
+
+# Properties of a Good Solution
+The solution we design should feature the following attributes:
+
+**Minimize Required `op-node` Diff by Block Builder Developers**. Developers of Block Builders should not have to modify `op-node` at all in order to integrate. They should be able to design their system at any level of complexity, from the outside.
+
+**Deny DOS Vectors**.
+The Sequencer should not be newly exposed to mechanisms which could be abused to reduce performance.
+
+**Support Fallback**.
+When Block Builders fail to deliver valid blocks, it should not impact the service liveness, and the "traditionally" built block should be usable in its place.
+
+**Be Asynchronous**.
+The Sequencer should not take Block Builders as a blocking source of latency.
+
+# Proposed Solution
+We should modify our Sequencing to support an event-driven, asyncrhonous mechanism for discovering and publishing blocks. These modifications should be codified as spec, and the API interfaces between the Sequencer and Builders should be especially defined.
+
+Recently, the Derivation package has been refactored to use an event-driven architecture. This work was defined in the previous "Derivers" [design document](https://github.com/ethereum-optimism/design-docs/pull/1).
+
+In this event-based architecture, the Sequencer regularly emits events to either:
+* Begin block building
+* Seal the block (end block building)
+* FCU the block (to make it cannonical against the Execution Engine)
+
+To support PBS, the Sequencer is newly responsible for managing the flow of a Block Builder as an outside agent. To accomplish this, it should emit timing-based calls which tell the block builder critical information needed to provide blocks, including block attributes and timing information.
+
+## High Level Overview
+Here is a diagram of the components and their interactions.
+```mermaid
+sequenceDiagram
+    participant Customer
+    participant ProxyD
+    participant Block Builder
+    participant Execution Engine
+    participant Sequencer
+    participant Replicas
+
+    Customer->>+ProxyD: Send Transaction
+    ProxyD->>+Block Builder: Forward Transaction
+    ProxyD->>+Execution Engine: Forward Transaction
+
+        Note over Sequencer: On a regular basis...
+    Sequencer->>Block Builder: Start Building <Attributes>
```

Comment:
I am not sure what makes the most sense here with the particular flow to be honest, i naively thought that the block builder would ask the sequencer for the same info that the sequencer pushes to the EL when it tells it to build a block, then the block builder would push a block to the sequencer and the sequencer would then have a choice to seal either the local block or the block builders block

Reply:
Yeah, all sorts of sync/async push/pull permutations are possible for sure, and none of them are outright wrong. However, I think this particular arrangement is good and I justify it as:

*Sequencer should Push Block Attributes to BBs* **because** the sequencer has the most timely information regarding a new block needing to be built. If a Block Builder was expected to pull attributes from the Sequencer, it might accidentally do it too late, wasting time at the start of the block-period.

*BBs should Push Payloads to the Sequencer* **because** the sequencer should not add extra delay in the creation of blocks waiting for the BB. If the BB delivers a payload in time, great. If the BB is unperformant, it should not impact block production. This also gives the BB the opportunity to send out blocks iteratively, or to send the block early if it is done building. We *could* make this a pull from the sequencer instead, doing the call in a separate thread, and abandoning if the payload doesn't come back, but seeing as we have this nice Event-Driven Derivers architecture, I wanted the builder to be able to trigger an event workflow within the sequencer to consider the BB payload.

The "Notification of Deadline" message is a timing mechanism meant to help synchronize the BB and keep it from spending too long building, only to find that the sequencer has moved on.

We could totally consider other arrangements of these responsibilities, I would just want to consider what advantages they give us.

----

> the sequencer would then have a choice to seal either the local block or the block builders block

This might be a gap on my understanding of block-sealing. It was my impression that the BB could seal blocks it creates, and therefore both the EE and BB would be sealing their own blocks. But now that you point it out, I bet that there's a signing authority thing that has to happen with blocks, and in that case: *yes*, we should rearrange this diagram somewhat. The BB and EE should both provide *unsealed* payloads to the Sequencer, and then the sequencer chooses which one gets sealed by the EE, and all other steps continue as normal.